refactor: Clean up messy strformat and bilingual_str usages

ryanofsky opened the pull request:
This PR cleans up string formatting in the codebase so other PRs adding compile time checking can be simpler and easier to review (specifically #30928, #31061, #31074, and #31149).

Currently these PRs are hard to review because in addition to changing formatting APIs, they have to update callers that are using the API's in unusual ways. Clean up these callers now so later PRs can be simpler. Specifically:

- Use string literals instead of `std::string` format strings to enable more compile-time checking.
- Avoid using untranslated bilingual strings as format strings. Use originals so they can by checked at compile time.
- Favor `Untranslated(strprintf(...))` over `strprintf(Untranslated(...), ...)` for consistency and to prevent translated and untranslated strings from being unintentionally combined.

ryanofsky committed:
refactor: Don't embed translated string in untranslated string.

This could produce an english error message containing non-english string
fragments if PopulateAndValidateSnapshot started returning any translated
strings in the future. This change is also needed to make the next
scripted-diff commit work.

ryanofsky committed:
scripted-diff: Replace strprintf(Untranslated) with Untranslated(strprintf)

This makes code more consistent and makes it easier to add compile-time checking to
enforce that format strings contain the right specifiers, because it stops
using Untranslated() to create the format string, so the Untranslated()
function will not need to get involved in formatting.

-BEGIN VERIFY SCRIPT-
quote='"[^"]+"'
quotes="(?:$quote|\\s)*"
nonparens="[^()]*"
single_level_paren="\($nonparens\)"
double_level_paren="\($nonparens\($nonparens\)$nonparens\)"
exprs="(?:$double_level_paren|$single_level_paren|$nonparens)*"
git grep -l 'Untranslated' | xargs perl -0777 -i -pe "s/strprintf\((\\W*)Untranslated\(($quotes)\)($exprs)(\))/Untranslated(\1strprintf(\2\3))/gs"
-END VERIFY SCRIPT-

maflcko reviewed (commented):
I don't mind the changes here, but I don't think the motivation is correct and it seems verbose and misleading.

I don't think it matters much whether this is merged or not, but you could just replace the motivation with:

"This replacement makes the implementation of some compile-time checks in the future easier. Also, it makes the code a bit more consistent in that `Untranslated` isn't allowed as a format-string anymore, only plain `std::string`s, or translated strings. The downside would be that developers may have to use `operator+` more often, making it a bit harder to write code involving formatting of (un)translated strings. However, if it becomes too annoying, the changes here should be easy to revert."

ryanofsky commented on src/init.cpp:
```diff
@@ -901,7 +901,9 @@ bool AppInitParameterInteraction(const ArgsManager& args)
     bilingual_str warnings;
     for (const auto& section : args.GetUnrecognizedSections()) {
-        warnings += strprintf(Untranslated("%s:%i ") + _("Section [%s] is not recognized.") + Untranslated("\n"), section.m_file, section.m_line, section.m_name);
+        warnings += Untranslated(strprintf("%s:%i ", section.m_file, section.m_line)) +
+                    strprintf(_("Section [%s] is not recognized."), section.m_name) +
```
> I still don't think this is correct.

What part of this is not correct? Is there a specific sentence or something?

> You are using `+` to concatenate `Untranslated` and `_()` (translated), possibly mixing English strings with non-English ones. Also, this compiles, so there are no compile-time checks preventing the "mix".

There is nothing in this commit message (19f49f138c4756b5219e715d8749d15be07e2c81) which mentions mixing English or non-english strings.

As a whole, this PR is updating around 30 strprintf calls to be more readable and consistent and to, yes, make 27 of them now trigger compiler errors if there is an attempt to pass a `bilingual_str` as a `strprintf` parameter, preventing accidental mixing English and non-english strings.

Obviously it should not be compile error to intentionally mix translated and non-translated strings, and this is still possible with `+`, which the code is now using in 3 more places.

ryanofsky committed:
refactor: Avoid concatenation of format strings

Instead just concatenate already formatted strings. This allows untranslated
format strings to be checked at compile time now, and translated format strings
to be checked at compile time in #31061.

ryanofsky committed:
refactor: Use + instead of strformat to concatenate translated & untranslated strings

This change manually removes two strprintf(Untranslated...) calls. All
remaining calls are removed in the next scripted-diff commit.

Removing these calls makes code more consistent and makes it easier to
implement compile-time checking enforcing that format strings contain valid
specifiers, by avoiding the need for the Untranslated() function to be involved
in formatting.

Additionally, using + and += instead of strprintf here makes code a little
shorter, and more type-safe because + unlike strprintf only works on strings of
the same type, making it less likely english strings and bilingual strings will
be unintentionally combined.

ryanofsky commented:
Updated 9a101d2928d14e39365064dc976615b4b9674ca5 -> 022ffd149a64f944d1ade75664e8dc19d04bcda7 ([`pr/bfmt.1`](https://github.com/ryanofsky/bitcoin/commits/pr/bfmt.1) -> [`pr/bfmt.2`](https://github.com/ryanofsky/bitcoin/commits/pr/bfmt.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bfmt.1-rebase..pr/bfmt.2)) to include more cleanups and help simplify implementation of compile-time translated string format checking in #31061 or #31074

---

re: https://github.com/bitcoin/bitcoin/pull/31072#pullrequestreview-2362888700

> I don't mind the changes here, but I don't think the motivation is correct and it seems verbose and misleading.

I don't know what was or misleading, but I rewrote description and will paste old text in comment below in case it's useful. Sorry for verbosity, I like to think by writing, but it doesn't always result in the most helpful thing to read thing to read.

maflcko reviewed (approved):
lgtm. Left some comments for context.

stickies-v commented on src/wallet/rpc/backup.cpp:
```diff
@@ -534,7 +534,7 @@ RPCHelpMan importwallet()
         // Use uiInterface.ShowProgress instead of pwallet.ShowProgress because pwallet.ShowProgress has a cancel button tied to AbortRescan which
         // we don't want for this progress bar showing the import progress. uiInterface.ShowProgress does not have a cancel button.
-        pwallet->chain().showProgress(strprintf("%s " + _("Importing…").translated, pwallet->GetDisplayName()), 0, false); // show progress dialog in GUI
+        pwallet->chain().showProgress(strprintf("%s %s", pwallet->GetDisplayName(), _("Importing…").translated), 0, false); // show progress dialog in GUI
```
> This commit (message) only becomes true after commit [4ce60d8](https://github.com/bitcoin/bitcoin/commit/4ce60d82a0c1e21ee7ecba69e4c8926720829429) from #30928. So I wonder if [4ce60d8](https://github.com/bitcoin/bitcoin/commit/4ce60d82a0c1e21ee7ecba69e4c8926720829429)+faa34ad0a7f5271958e5313e7aae2baaed41587e could be split up into a first pull request, as everything else is based on it?

This is now ~done in https://github.com/bitcoin/bitcoin/pull/31149

hebasto commented:
Concept ACK.

ryanofsky commented on src/validation.cpp:
```diff
@@ -6091,7 +6091,7 @@ SnapshotCompletionResult ChainstateManager::MaybeCompleteSnapshotValidation()
         auto rename_result = m_snapshot_chainstate->InvalidateCoinsDBOnDisk();
         if (!rename_result) {
-            user_error = strprintf(Untranslated("%s\n%s"), user_error, util::ErrorString(rename_result));
+            user_error += Untranslated("\n") + util::ErrorString(rename_result);
```
re: https://github.com/bitcoin/bitcoin/pull/31072#discussion_r1801118756

Thanks, we also talked about this offline and I agree that strprintf can often be nicer than concatenation for building strings, but I think this approach doesn't work very well for untranslated bilingual strings. For example in the HTML case, you would probably only want to format the translated strings as HTML, not both translated and original strings. I think the downsides of supporting untranslated format strings outweigh the upsides. (Downsides being having two inconsistent ways of formatting the same strings, having a more complicated interface, and because of the potential for introducing mixed translation bugs.)

ryanofsky reviewed (commented):
Rebased 022ffd149a64f944d1ade75664e8dc19d04bcda7 -> 46c55ab427c0771cacaed7a20612ffe2cc3543db ([`pr/bfmt.2`](https://github.com/ryanofsky/bitcoin/commits/pr/bfmt.2) -> [`pr/bfmt.3`](https://github.com/ryanofsky/bitcoin/commits/pr/bfmt.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bfmt.2-rebase..pr/bfmt.3)) due to conflict with #31042, also addressing review comments.

ryanofsky commented:
Rebased 1e9ea7de0b5ee6d3179930d4fc93c8e276e525bd -> 85df2fbf26c73f97f85797868155247c11a4ccd6 ([`pr/bfmt.4`](https://github.com/ryanofsky/bitcoin/commits/pr/bfmt.4) -> [`pr/bfmt.5`](https://github.com/ryanofsky/bitcoin/commits/pr/bfmt.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bfmt.4-rebase..pr/bfmt.5)) after #31287 (just for clarity, there were no merge conflicts)

l0rinc commented on test/lint/run-lint-format-strings.py:
```diff
@@ -13,7 +13,7 @@
 
 FALSE_POSITIVES = [
-    ("src/clientversion.cpp", "strprintf(_(COPYRIGHT_HOLDERS).translated, COPYRIGHT_HOLDERS_SUBSTITUTION)"),
+    ("src/clientversion.cpp", "strprintf(_(COPYRIGHT_HOLDERS), COPYRIGHT_HOLDERS_SUBSTITUTION)"),
```
what is the reason for these still being false positives?

l0rinc commented on src/wallet/feebumper.cpp:
```diff
@@ -24,24 +24,24 @@ namespace wallet {
 {
     if (wallet.HasWalletSpend(wtx.tx)) {
-        errors.push_back(Untranslated("Transaction has descendants in the wallet"));
+        errors.emplace_back(Untranslated("Transaction has descendants in the wallet"));
```
nit: it's simple enough to review, but if you edit again consider a scripted diff

l0rinc reviewed (approved):
ACK 85df2fbf26c73f97f85797868155247c11a4ccd6

I have basically redone the change locally to understand it better and mostly left questions or nits, I'm fine with merging as is

ryanofsky commented on src/clientversion.cpp:
```diff
@@ -71,7 +71,7 @@ std::string FormatSubVersion(const std::string& name, int nClientVersion, const
 std::string CopyrightHolders(const std::string& strPrefix)
 {
-    const auto copyright_devs = strprintf(_(COPYRIGHT_HOLDERS).translated, COPYRIGHT_HOLDERS_SUBSTITUTION);
+    const auto copyright_devs = strprintf(_(COPYRIGHT_HOLDERS), COPYRIGHT_HOLDERS_SUBSTITUTION).translated;
```
re: https://github.com/bitcoin/bitcoin/pull/31072#discussion_r1845404225

Right, I think it would be bad if changing or tweaking the name of the program required newly translating "The %s developers" with the new name of the program hardcoded.

ryanofsky reviewed (commented):
Thanks for the reviews!

Updated 85df2fbf26c73f97f85797868155247c11a4ccd6 -> 3b24c24889c09f6c8f33408b57c2042c4e565388 ([`pr/bfmt.5`](https://github.com/ryanofsky/bitcoin/commits/pr/bfmt.5) -> [`pr/bfmt.6`](https://github.com/ryanofsky/bitcoin/commits/pr/bfmt.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bfmt.5..pr/bfmt.6)) fixing a commit description (no code changes)

l0rinc commented:
ACK 3b24c24889c09f6c8f33408b57c2042c4e565388

ryanofsky commented:
Rebased 3b24c24889c09f6c8f33408b57c2042c4e565388 -> 0184d33b3d28fe78a2ee228417ebfd6f46fcaee5 ([`pr/bfmt.6`](https://github.com/ryanofsky/bitcoin/commits/pr/bfmt.6) -> [`pr/bfmt.7`](https://github.com/ryanofsky/bitcoin/commits/pr/bfmt.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bfmt.6-rebase..pr/bfmt.7)) with no changes to simplify after #31295

l0rinc commented:
ACK 0184d33b3d28fe78a2ee228417ebfd6f46fcaee5 - no overall difference because of the rebase

hodlinator commented on src/wallet/feebumper.cpp:
```diff
@@ -74,10 +74,10 @@ static feebumper::Result CheckFeeRate(const CWallet& wallet, const CMutableTrans
     CFeeRate minMempoolFeeRate = wallet.chain().mempoolMinFee();
 
     if (newFeerate.GetFeePerK() < minMempoolFeeRate.GetFeePerK()) {
-        errors.push_back(strprintf(
-            Untranslated("New fee rate (%s) is lower than the minimum fee rate (%s) to get into the mempool -- "),
```
A bit curious about why the caliber of the scripted diff was so powerful, but I guess this multi-line case is it. Would also catch potential other cases like it which could have been merged between the parent commit of this PR and when it ended up being merged.

hodlinator reviewed (commented):
post-merge-ACK

Really cleaned the slate nicely for #31061.
